src/python/pants/core/util_rules/environments.py

```diff
@@ -383,30 +384,55 @@ async def _warn_on_non_local_environments(specified_targets: Iterable[Target], s
     """
 
     env_names = [
-        (target, target[EnvironmentField].value)
+        (target[EnvironmentField].value, target)
         for target in specified_targets
         if target.has_field(EnvironmentField)
     ]
 
+    env_names_and_targets = [
+        (i[0], tuple(j[1] for j in i[1])) for i in groupby(sorted(env_names), lambda x: x[0])
```

Comment:
Nice change!

I think this could be slightly more readable though:
```python
env_names_and_targets = [
    (env_name, tuple(target for _, target in group)) 
    for env_name, group in groupby(sorted(env_names), lambda x: x[0])
]        
```